lib/maths/unittest/COutliersTest.cc

```diff
@@ -389,6 +392,100 @@ void COutliersTest::testEnsemble() {
     }
 }
 
+void COutliersTest::testFeatureInfluences() {
+
+    // Test calculation of outlier significant features.
+
+    // We have the following basic geometry:
+    //   1) Two clusters of points (x, y)
+    //   2) Three outliers
+    //         i) One displaced in the x-direction
+    //        ii) One displaced in the y-direction
+    //       iii) One displaced in both directions
+
+    TFactoryFunc toMainMemoryDataFrame{[](const TPointVec& points) {
+        return test::CDataFrameTestUtils::toMainMemoryDataFrame(points);
+    }};
+    TFactoryFunc toOnDiskDataFrame{[](const TPointVec& points) {
+        return test::CDataFrameTestUtils::toOnDiskDataFrame(
+            boost::filesystem::current_path().string(), points);
+    }};
+    TFactoryFunc factories[]{toMainMemoryDataFrame, toOnDiskDataFrame};
+    std::size_t numberPartitions[]{1, 3};
+    std::size_t numberThreads[]{1, 4};
+
+    test::CRandomNumbers rng;
+
+    TDoubleVec means[]{{0.0, 0.0}, {100.0, 100.0}};
+    TDoubleVecVec covariances[]{{{7.0, 1.0}, {1.0, 8.0}}, {{5.0, 2.0}, {2.0, 12.0}}};
+
+    TPointVec points;
+    for (std::size_t i = 0; i < 2; ++i) {
+        TDoubleVecVec inliers;
+        rng.generateMultivariateNormalSamples(means[i], covariances[i], 100, inliers);
+
+        points.resize(points.size() + inliers.size(), TPoint{2});
+        for (std::size_t j = inliers.size(); j > 0; --j) {
+            for (std::size_t k = 0; k < inliers[j - 1].size(); ++k) {
+                points[points.size() - j](k) = inliers[j - 1][k];
+            }
+        }
+    }
+    points.emplace_back(2);
+    points.back() << 0.0, 50.0;
+    points.emplace_back(2);
+    points.back() << 150.0, 100.0;
+    points.emplace_back(2);
+    points.back() << -30.0, -30.0;
+
+    std::size_t outlierIndexes[]{points.size() - 3, points.size() - 2, points.size() - 1};
+
+    core::stopDefaultAsyncExecutor();
+
+    std::string tags[]{"sequential", "parallel"};
+
+    // Test in/out of core.
+    for (std::size_t i = 0; i < 2; ++i) {
+
+        // Test sequential then parallel.
+        for (std::size_t j = 0; j < 2; ++j) {
+            LOG_DEBUG(<< "Testing " << tags[j]);
+
+            auto frame = factories[i](points);
+            maths::COutliers::SComputeParameters params{
+                numberThreads[j], numberPartitions[i], true, 0.05};
+            maths::COutliers::compute(params, *frame);
+
+            bool passed{true};
+            frame->readRows(1, [&](core::CDataFrame::TRowItr beginRows,
+                                   core::CDataFrame::TRowItr endRows) {
+                for (auto row = beginRows; row != endRows; ++row) {
+                    if (row->index() == outlierIndexes[0]) {
+                        LOG_DEBUG(<< "x-significance = " << (*row)[3]
+                                  << ", y-significance = " << (*row)[4]);
+                        passed &= (1.0 - (*row)[4] < 0.005);
```

Comment:
Should we also check the influence of the feature which is not outlying is a low score?

Reply:
Currently, I don't actually control the feature influence of inliers.

Given the nature of the calculation these could still be dominated by one feature whose influence is close to 1. This is because we're computing relative influence so renormalise by the sum of individual feature influences. The intention is we don't output the feature influence at all if the outlier score is low.

I could rejig the calculation to impose this constraint, but my inclination is to leave it. As a result this sort of assertion is not possible.

Reply:
That said, it may be possible to make a meaningful assertion about the average behaviour of inlier influences. I'll ponder

Reply:
Discussed this further with @dimitris-athanasiou, his suggestion was to check the other coordinates in these cases, which is sensible. In fact, we can do even better and check that the scores are normalised, which I do as part of [a4f6481](https://github.com/elastic/ml-cpp/pull/405/commits/a4f6481a96f2db3041c5cd48c68d45366d47a32d). This also checks to see that the influences don't show evidence of bias.